File: 1466.reorder-routes-to-make-all-paths-lead-to-the-city-zero.py
#
# @lc app=leetcode id=1466 lang=python3
#
# [1466] Reorder Routes to Make All Paths Lead to the City Zero
#

from typing import List
from collections import defaultdict, deque


# @lc code=start
class Solution:
    def minReorder(self, n: int, connections: List[List[int]]) -> int:
        graph = defaultdict(list)

        for u, v in connections:
            graph[u].append((v, 1))
            graph[v].append((u, 0))

        visited = set()
        queue = deque([0])
        visited.add(0)
        changes = 0

        while queue:
            city = queue.popleft()

            for neighbor, cost in graph[city]:
                if neighbor not in visited:
                    visited.add(neighbor)
                    queue.append(neighbor)
                    changes += cost

        return changes


# @lc code=end

File: 1466.reorder-routes-to-make-all-paths-lead-to-the-city-zero.go
/*
 * @lc app=leetcode id=1466 lang=golang
 *
 * [1466] Reorder Routes to Make All Paths Lead to the City Zero
 */
package leetcode

// @lc code=start
func dfs(cur, parent int, graph [][][]int, total *int) {
	for _, edge := range graph[cur] {
		next := edge[0]
		cost := edge[1]

		if next == parent {
			continue
		}

		*total += cost
		dfs(next, cur, graph, total)
	}
}

func minReorder(n int, connections [][]int) int {
	graph := make([][][]int, n)
	for _, pairs := range connections {
		u := pairs[0]
		v := pairs[1]

		graph[u] = append(graph[u], []int{v, 1})
		graph[v] = append(graph[v], []int{u, 0})
	}

	total := 0
	dfs(0, -1, graph, &total)

	return total
}

// @lc code=end
